File: Algorithms/Python/Balanced-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):

    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        def getDepth(node, depth):
            if not node:
                return depth
            depth += 1
            return max(getDepth(node.left, depth), getDepth(node.right, depth))

        def comparelr(nodeA, nodeB):
            if abs(getDepth(nodeA, 0) - getDepth(nodeB, 0)) > 1:
                return False
            a, b = True, True
            if nodeA:
                a = comparelr(nodeA.left, nodeA.right)
            if nodeB:
                b = comparelr(nodeB.left, nodeB.right)
            return a and b

        if not root:
            return True
        return comparelr(root.left, root.right)

File: Algorithms/Java/Balanced Binary Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
		if (root == null)
			return true;
		int a = maxDepth(root.left);
		int b = maxDepth(root.right);
		if (Math.abs(a - b) > 1)
			return false;
		else
			return isBalanced(root.left) && isBalanced(root.right);
	}

	public int maxDepth(TreeNode root) {
		if (root == null)
			return 0;
		return find(root, 1);
	}

	public int find(TreeNode tree, int n) {
		if (tree.left != null && tree.right != null)
			return Math.max(find(tree.left, n + 1), find(tree.right, n + 1));
		else if (tree.left != null)
			return find(tree.left, n + 1);
		else if (tree.right != null)
			return find(tree.right, n + 1);
		else
			return n;
	}
}